src/main/java/org/minions/devfund/richard/battleship/Ship.java

```diff
@@ -0,0 +1,156 @@
+package org.minions.devfund.richard.battleship;
+
+import java.util.stream.IntStream;
+
+/**
+ * Class.
+ */
+public abstract class Ship {
+
+    private static final String SUNK_CHAR = "x";
+    private static final String NOT_SUNK_CHAR = "S";
+
+    private int bowRow;
+    private int bowColumn;
+    private int length;
+    private boolean horizontal;
+    private boolean[] hit;
+    private int number;
+
+    /**
+     * @return String.
+     */
+    abstract String getShipType();
+
+    /**
+     * @param row        row.
+     * @param column     column.
+     * @param horizontal horizontal.
+     * @param ocean      ocean.
+     * @return boolean.
+     */
+    boolean okToPlaceShipAt(final int row, final int column, final boolean horizontal, final Ocean ocean) {
+        this.bowRow = row;
+        this.bowColumn = column;
+        this.horizontal = horizontal;
+        return ShipHelper.isValidPosition(this, ocean);
+    }
+
+    /**
+     * @param row        row.
+     * @param column     column.
+     * @param horizontal horizontal.
+     * @param ocean      ocean.
+     */
+    void placeShipAt(final int row, final int column, final boolean horizontal, final Ocean ocean) {
+        this.bowRow = row;
+        this.bowColumn = column;
+        this.horizontal = horizontal;
+        ShipHelper.setShip(this, ocean);
+    }
+
+    /**
+     * @param row    row.
+     * @param column column.
+     * @return boolean.
+     */
+    boolean shootAt(int row, int column) {
+        boolean flag = false;
+        if (!isSunk()) {
+            if (!horizontal && column == bowColumn && row >= bowRow && row < bowRow + length) {
+                hit[row - bowRow] = true;
+                flag = true;
+            } else if (horizontal && row == bowRow && column >= bowColumn && column < bowColumn + length) {
+                hit[column - bowColumn] = true;
+                flag = true;
+            }
+        }
+        return flag;
```

Comment:
This may be refactored like this.
```java
    boolean shootAt(int row, int column) {
        if (isSunk()) {
            return false;
        }
        final boolean isHittable = horizontal ? row == bowRow && column >= bowColumn && column < bowColumn + length
                : column == bowColumn && row >= bowRow && row < bowRow + length;
        if (isHittable) {
            final int hitPosition = horizontal ? column - bowColumn : row - bowRow;
            hit[hitPosition] = true;
            return true;
        }
        return false;
    }
```